kubelet: Record a metric for latency of pod status update

smarterclayton opened the pull request:
Track how long it takes for pod updates to propagate from detection to successful change on API server. Will guide future improvements in pod start and shutdown latency. Currently pod status updates have been observed to take 30s or more to propagate to the apiserver even on underutilized nodes, and this metric will help quantify the level of latency for status syncs and help identify the need for improvements.

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A new `pod_status_sync_duration_seconds` histogram is reported at alpha metrics stability that estimates how long the Kubelet takes to write a pod status change once it is detected.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 commented:
/remove-area kubeadm

binacs reviewed (approved):
/retest

ehashman commented:
/assign @dashpole 
for instrumentation

/triage accepted
/priority backlog

/remove-kind api-change bug
/kind feature

dashpole commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -634,6 +646,15 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
+	} else {
+		duration = time.Now().Sub(status.at).Truncate(time.Millisecond)
+	}
+	metrics.PodStatusSyncDuration.WithLabelValues(strconv.Itoa(0)).Observe(duration.Seconds())
```
Should we only observe a duration if status.at was non-zero?

smarterclayton replied:
Good question, it's typically an indicator that something unexpected happened, but it would skew the numbers.  Let me review.

smarterclayton replied:
I'll only observe if it's set, and then verify there are no wierd cases where a synthetic status is created that somehow escapes time (can't be 100% sure, but log observation in runs should be sufficient).

dashpole commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -482,6 +487,13 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
+	if cachedStatus.at.IsZero() {
+		newStatus.at = time.Now()
+	} else {
+		newStatus.at = cachedStatus.at
```
This essentially ensures we measure the longest outstanding status update for a pod, right?  E.g. if multiple status updates end up batched, we want to measure how long it took from the first update.

Can you add a brief comment?

smarterclayton replied:
Will do

dashpole reviewed (commented):
lgtm overall.

smarterclayton commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -482,6 +487,13 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 		podNamespace: pod.Namespace,
 	}
+
+	if cachedStatus.at.IsZero() {
```
Correct, which I'm not actually sure can happen except during some sort of invalidation.  Will look though.

smarterclayton replied:
Also during Kubelet restarts we will have a clean cache in status manager until every pod worker has started up and invoked SetPodStatus at least once.

smarterclayton commented:
/remove-lifecycle stale

ritazh commented:
/remove-sig auth
Feel free to add us back if needed.

smarterclayton committed:
kubelet: Record a metric for latency of pod status update

Track how long it takes for pod updates to propagate from detection
to successful change on API server. Will guide future improvements
in pod start and shutdown latency.

Metric is `kubelet_pod_status_sync_duration_seconds` and is ALPHA
stability. Histogram buckets are chosen based on distribution of
observed status delays in practice.

smarterclayton commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -155,6 +156,18 @@ var (
+			Buckets:        []float64{0.010, 0.050, 0.100, 0.500, 1, 5, 10, 20, 30, 45, 60},
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"priority"},
```
Priority was a follow on concept I was testing (some status updates like transition from unready to ready or running to failed are "more important" than updates like a container message being propagated).

I'll remove it from this PR and introduce it as a separate commit in a subsequent PR.

smarterclayton commented:
Ok, I think this is ready for re-review with all comments addressed, PTAL when you have the chance.

dashpole reviewed (approved):
/lgtm
